### PR TITLE
Integrate with COS for metrics

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

Integrate microk8s charm with grafana-agent.

**WARNING**: This is waiting on a few PRs to be merged and fixed on COS for the integration to work correctly

### CODE REVIEW

**Comment by jnsgruk** [comment]
Great to see this! I'd be interested to know if this could be simplified by the prometheus charm lib -- @sed-i might be able to shed some light, or at least understand why the lib wasn't sufficient :)

File: src/metrics.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        "tls_config": {
+            "insecure_skip_verify": True,
+        },
+        "authorization": {
+            "credentials": token,
+        },
+    }
+
+    # apiserver, kube-scheduler, kube-controller-manager, kube-proxy
+    for job_name, targets in [
```
Comment (sed-i) [inline src/metrics.py:96]: This seems odd at first glance.
Would you mind also sharing the output of the prometheus `/api/v1/targets` endpoint?

**Review by sed-i — COMMENTED** [review]
Left a few questions.

But I'm confused about the topology. Is this the intended use-case:

```mermaid
graph LR

subgraph lxd-controller
charmed-microk8s
end

subgraph k8s-controller
prometheus-k8s
end

charmed-microk8s ---|prometheus_scrape| prometheus-k8s
```

Not sure how it can work this way at all :thinking:

File: src/metrics.py (hunk 1 above)
Comment (neoaggelos) [reply src/metrics.py]: Alerts and dashboards (to be added) for individual components often look like `$metric{job="kubelet"} >= $expr` for the different components, thus we need to add the `job="$component" labels during scrape.

File: src/charm.py [hunk 2]
```diff
@@ -291,6 +312,56 @@ def add_node(self, event: RelationJoinedEvent):
...
+        scrape_configs = metrics.build_scrape_configs(token, control_plane_nodes, worker_nodes)
+        for relation in self.model.relations["metrics"]:
+            relation.data[self.app]["scrape_jobs"] = json.dumps(scrape_configs)
+            relation.data[self.app]["scrape_metadata"] = json.dumps(
+                {
+                    "model": self.model.name,
+                    "model_uuid": self.model.uuid,
+                    "application": self.app.name,
+                }
+            )
```
[the remark this answers is not in the document]
Comment (neoaggelos) [reply src/charm.py]: Here's a branch that uses `MetricsEndpointProvider` and `JujuTopology` for metrics instead https://github.com/canonical/charm-microk8s/commit/03fa095a2690d0e39023060330aa608ca8e2e44f

Some drawbacks/feedback:
- vendoring both libraries, and having to maintain newer versions for them for very little gain
- blocking bugs in the libraries (https://github.com/canonical/prometheus-k8s-operator/pull/490) and (https://github.com/canonical/prometheus-k8s-operator/pull/489), though these were both quite easy to fix after spending some time debugging
- we need per-unit labels, which is not too easy as of now (perhaps could be supported via a `prometheus_scrape_unit_labels` per-unit config?) for us, we need to set the `node` on each unit to the kubernetes node name. we are working around this at a moment by using the unit name instead, which will hopefully not cause problems down the line
- core logic is dependent on the libraries and is therefore not unit tested adequately. if we added the extra checks, we would be effectively unit testing the libraries themselves instead.

Happy to drive the discussion forward on this (i have already created 2 PRs that improve things for us), and we could consider using the libraries in the future, but we have quite a lot on our plate at the moment.

File: src/cos_agent.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#
+# Copyright 2023 Canonical, Ltd.
+#
+from charms.grafana_agent.v0.cos_agent import COSAgentProvider
+
+
+class Provider(COSAgentProvider):
+    @property
+    def _scrape_jobs(self):
+        return self._metrics_endpoints()
```
Comment (neoaggelos) [inline src/cos_agent.py:10]: May be dropped after https://github.com/canonical/grafana-agent-k8s-operator/pull/223 is complete

**Comment by neoaggelos** [comment]
This is currently blocked until the following upstream issues and PRs have been addressed:

- https://github.com/canonical/grafana-agent-k8s-operator/pull/219
- https://github.com/canonical/grafana-agent-k8s-operator/pull/220
- https://github.com/canonical/grafana-agent-k8s-operator/pull/221
- https://github.com/canonical/grafana-agent-k8s-operator/pull/222
- https://github.com/canonical/grafana-agent-k8s-operator/pull/223
- https://github.com/simskij/grafana-agent/issues/22
- https://github.com/canonical/grafana-agent-k8s-operator/issues/190

File: tests/integration/test_observability.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    await e.model.deploy(
+        config.MK8S_GRAFANA_AGENT_CHARM,
+        channel=config.MK8S_GRAFANA_AGENT_CHANNEL,
+        application_name="grafana-agent",
+    )
+    await e.model.relate("microk8s", "grafana-agent")
+
+    # TODO(neoaggelos): enable tests after required issues are fixed
+    # - https://github.com/simskij/grafana-agent/issues/22
```
Comment (ktsakalozos) [inline tests/integration/test_observability.py:29]: Why is this issue against a user's repo? I see it closed, but is it addressed on the official grafana-agent repo?

File: tests/integration/test_observability.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    # - https://github.com/canonical/grafana-agent-k8s-operator/issues/190
```
Comment (ktsakalozos) [inline tests/integration/test_observability.py:30]: This is the only one still open, right?

File: src/hack/update_libs.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+#
```
Comment (ktsakalozos) [inline src/hack/update_libs.py:1]: What is this file for?

File: src/hack/update_alert_rules.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+#
```
Comment (ktsakalozos) [inline src/hack/update_alert_rules.py:1]: I think we are missing a docs page in this repo describing why we need these scripts, when should we call them and how.

File: src/hack/update_alert_rules.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import os
+import shutil
+import sys
+from datetime import datetime
+from urllib.request import urlopen
+
+import yaml
+
+VERSION = "v0.12.0"
```
Comment (ktsakalozos) [inline src/hack/update_alert_rules.py:12]: How do we know this is the right version?

File: src/hack/update_kube_state_metrics.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#
+# Copyright 2023 Canonical, Ltd.
+#
+import sys
+from datetime import datetime
+from urllib.request import urlopen
+
+VERSION = "v2.9.2"
```
Comment (ktsakalozos) [inline src/hack/update_kube_state_metrics.py:8]: Same here, when should we update this version?

File: docs/architecture.md [hunk 10]
```diff
@@ -65,10 +65,22 @@ charm-microk8s:                     # Root directory
 - docs:
   - architecture.md                 # Document architecture decisions for the charm
   - development.md                  # Getting started with developing the charm and running tests
+- lib:
```
Comment (ktsakalozos) [inline docs/architecture.md:68]: I feel that this list of files is doomed to be stale. Maybe we should stop at the level of directories and not files.

File: docs/observability.md [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+### Deploy MicroK8s
+
+```bash
+# Add a machine model and deploy MicroK8s
+juju add-model microk8s
+juju deploy microk8s --channel edge --config hostpath_storage=true --constraints 'mem=8G'
+juju expose microk8s
+
+# Add MicroK8s as Kubernetes cloud to Juju
+juju run --unit microk8s/leader -- microk8s config | juju add-k8s k8s --controller "$(juju controller-config controller-name)"
```
Comment (ktsakalozos) [inline docs/observability.md:16]: You may want to get the IP of the API endpoint from juju here. This is needed for example on AWS clouds.

File: docs/observability.md [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+# Integrate MicroK8s with grafana-agent
+juju relate microk8s grafana-agent
+```
+
+### Access Grafana
+
+Retrieve the admin password for Grafana using:
+
+```bash
+juju run-action --wait -m observability grafana/0 get-admin-password
```
Comment (ktsakalozos) [inline docs/observability.md:61]: Should this be just `juju run` ?
```
jackal@aurora:~/workspace/microk8s-core-addons$ juju version
3.1.5-genericlinux-amd64
jackal@aurora:~/workspace/microk8s-core-addons$ juju run-action
ERROR juju: "run-action" is not a juju command. See "juju --help".
```

File: docs/observability.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+```
+
+Then, if LoadBalancer IP is not reachable from your host, use either `sshuttle` or `ssh`
```
Comment (ktsakalozos) [inline docs/observability.md:64]: I used kubectl port-forward. Maybe it is more complex.
```
kubectl port-forward -n mkcos service/grafana 10443:3000
```

File: docs/observability.md [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+```bash
+# Access Grafana at http://localhost:8000/observability-grafana/
+juju ssh -m microk8s microk8s/leader -L 8000:10.43.5.10:80
```
Comment (ktsakalozos) [inline docs/observability.md:68]: Should we reduce the range of IPs metallb has?

File: docs/observability.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+# Access Grafana at http://10.43.5.10/observability-grafana/
+sshuttle -r $microk8snodeip 10.43.5.10/32
+```
+
```
Comment (ktsakalozos) [inline docs/observability.md:73]: Please add a line here saying: `Now point your browser to ....`

File: docs/observability.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+The charm also automatically deploys [`kube-state-metrics`](https://github.com/kubernetes/kube-state-metrics) to the cluster. The manifests for `kube-state-metrics` can be found in [src/deploy/kube-state-metrics.yaml](../src/deploy/kube-state-metrics.yaml) and can be automatically updated using the [src/hack/update_kube_state_metrics.py](../src/hack/update_kube_state_metrics.py) script.
+
+### Implementation Notes and reference
+
+#### Update manifests from upstream projects
+
+To update upstream manifests, use the scripts in `src/hack`.
+
+1. Check the latest upstream version of each component (and ensure that it is compatible with the MicroK8s version we deploy)
```
Comment (ktsakalozos) [inline docs/observability.md:92]: I feel there more to this. Could we be more precise which versions of what components should match/checked for compatibility? What are the `each component`?

File: docs/observability.md (hunk 11 above)
Comment (neoaggelos) [reply docs/observability.md:16]: This kubeconfig will contain the internal IP of the microk8s node, which _should_ be reachable from the Juju controller instance in all cases.

File: docs/observability.md (hunk 14 above)
Comment (neoaggelos) [reply docs/observability.md:68]: This just shows an example range with internal IPs. It is up to the user to change this according to the deployment requirements.

**Pull request merged**